File: Utility/FocusAttacher.vb
Imports System.Windows

Public NotInheritable Class FocusAttacher

    Private Sub New()
    End Sub

    Public Shared ReadOnly FocusProperty As DependencyProperty = DependencyProperty.RegisterAttached("Focus", GetType(Boolean), GetType(FocusAttacher), New PropertyMetadata(False, New PropertyChangedCallback(AddressOf FocusChanged)))

    Public Shared Function GetFocus(ByVal d As DependencyObject) As Boolean
        Return CBool(d.GetValue(FocusProperty))
    End Function

    Public Shared Sub SetFocus(ByVal d As DependencyObject, ByVal value As Boolean)
        d.SetValue(FocusProperty, value)
    End Sub

    Private Shared Sub FocusChanged(ByVal sender As Object, ByVal e As DependencyPropertyChangedEventArgs)
        If CBool(e.NewValue) Then
            DirectCast(sender, UIElement).Focus()
        End If
    End Sub

End Class

File: Utility/FocusAttacher.cs
using System;
using System.Windows;

namespace Utility
{
    public static class FocusAttacher
    {
        public static readonly DependencyProperty FocusProperty = DependencyProperty.RegisterAttached(
            "Focus",
            typeof (bool),
            typeof (FocusAttacher),
            new PropertyMetadata(false, FocusChanged));

        public static bool GetFocus(DependencyObject d)
        {
            return Convert.ToBoolean(d.GetValue(FocusProperty));
        }

        public static void SetFocus(DependencyObject d, bool value)
        {
            d.SetValue(FocusProperty, value);
        }

        private static void FocusChanged(object sender, DependencyPropertyChangedEventArgs e)
        {
            if (Convert.ToBoolean(e.NewValue))
            {
                ((UIElement) sender).Focus();
            }
        }
    }
}